File: src/lib/api/briefings.ts
import type { CaseCard } from '@/types';
import { type CaseDto, toCaseCard } from './cases';
import { ApiError, apiFetch } from './client';

// GET /api/v1/briefings/latest 응답 DTO(backend/app/schemas/briefing.py) — cases는 cases.ts의
// CaseDto/toCaseCard를 그대로 재사용한다(백엔드가 이미 CaseOut을 재사용하는 것과 대칭).
export interface BriefingDto {
  id: string;
  briefing_date: string;
  generated_at: string;
  cases: CaseDto[];
}

export interface Briefing {
  briefingId: string;
  briefingDate: string;
  generatedAt: string;
  // briefing_items rank 순으로 서버가 이미 정렬해 내려준다.
  cases: CaseCard[];
}

function toBriefing(dto: BriefingDto): Briefing {
  return {
    briefingId: dto.id,
    briefingDate: dto.briefing_date,
    generatedAt: dto.generated_at,
    cases: dto.cases.map(toCaseCard),
  };
}

// 아직 그날의 브리핑이 생성되지 않은 경우 백엔드가 404를 반환한다 — 예외로 던지지 않고 null로
// 변환해, 호출부가 "브리핑 없음"과 "요청 실패"를 구분해 처리하지 않아도 되게 한다(빈 상태 화면).
export async function fetchLatestBriefing(): Promise<Briefing | null> {
  try {
    return toBriefing(await apiFetch<BriefingDto>('/api/v1/briefings/latest'));
  } catch (err) {
    if (err instanceof ApiError && err.status === 404) return null;
    throw err;
  }
}

File: backend/app/services/briefings.py
"""briefings 도메인 읽기 서비스 — GET /api/v1/briefings/latest(R2.3). docs/DB_SCHEMA.md §4.9.

케이스 조립은 app.services.cases.get_case_out을 그대로 재사용한다 — 다시 만들지 않는다.
"""

from __future__ import annotations

from sqlalchemy import select
from sqlalchemy.orm import Session

from app.models.briefing import Briefing, BriefingItem
from app.models.case import Case
from app.schemas.briefing import BriefingOut
from app.services.cases import get_case_out


def get_latest_briefing_out(db: Session, company_id: str) -> BriefingOut | None:
    """company_id로 스코프된 최신 브리핑(briefing_date desc, generated_at desc) 1건을
    briefing_items(rank asc) 순서 그대로 케이스와 조립해 반환한다. 브리핑이 없으면 None."""
    briefing = db.execute(
        select(Briefing)
        .where(Briefing.company_id == company_id)
        .order_by(Briefing.briefing_date.desc(), Briefing.generated_at.desc())
        .limit(1)
    ).scalar_one_or_none()
    if briefing is None:
        return None

    items = db.execute(
        select(BriefingItem)
        .where(BriefingItem.company_id == company_id, BriefingItem.briefing_id == briefing.id)
        .order_by(BriefingItem.rank.asc())
    ).scalars().all()

    cases = []
    for item in items:
        case = db.execute(
            select(Case).where(Case.company_id == company_id, Case.id == item.case_id)
        ).scalar_one_or_none()
        if case is None:
            continue
        cases.append(get_case_out(db, company_id, case))

    return BriefingOut(
        id=briefing.id,
        briefing_date=briefing.briefing_date,
        generated_at=briefing.generated_at,
        cases=cases,
    )
